packages/postgres-database/src/simcore_postgres_database/utils_folders.py

```diff
@@ -0,0 +1,919 @@
+    UPDATE_FOLDER: ClassVar[_FolderPermissions] = _make_permissions(d=True)
+    DELETE_FOLDER: ClassVar[_FolderPermissions] = _make_permissions(d=True)
+    REMOVE_PROJECT_FROM_FOLDER: ClassVar[_FolderPermissions] = _make_permissions(d=True)
+
+    _MOVE_FOLDER_SOURCE: ClassVar[_FolderPermissions] = _make_permissions(
+        d=True,
+        description="apply to folder providing the data",
+    )
+    _MOVE_FOLDER_DESTINATION: ClassVar[_FolderPermissions] = _make_permissions(
+        w=True, description="apply on the folder receiving the data"
+    )
+    MOVE_FOLDER: ClassVar[_FolderPermissions] = _or_dicts_list(
+        [_MOVE_FOLDER_SOURCE, _MOVE_FOLDER_DESTINATION]
+    )
+
+
+NO_ACCESS_PERMISSIONS: _FolderPermissions = _make_permissions()
+
+VIEWER_PERMISSIONS: _FolderPermissions = _or_dicts_list(
+    [
+        _BasePermissions.LIST_FOLDERS,
+    ]
+)
+EDITOR_PERMISSIONS: _FolderPermissions = _or_dicts_list(
+    [
+        VIEWER_PERMISSIONS,
+        _BasePermissions.CREATE_FOLDER,
+        _BasePermissions.ADD_PROJECT_TO_FOLDER,
+    ]
+)
+OWNER_PERMISSIONS: _FolderPermissions = _or_dicts_list(
+    [
+        EDITOR_PERMISSIONS,
+        _BasePermissions.SHARE_FOLDER,
+        _BasePermissions.UPDATE_FOLDER,
+        _BasePermissions.DELETE_FOLDER,
+        _BasePermissions.REMOVE_PROJECT_FROM_FOLDER,
+        _BasePermissions.MOVE_FOLDER,
+    ]
+)
+
+_ROLE_TO_PERMISSIONS: dict[FolderAccessRole, _FolderPermissions] = {
+    FolderAccessRole.NO_ACCESS: NO_ACCESS_PERMISSIONS,
+    FolderAccessRole.VIEWER: VIEWER_PERMISSIONS,
+    FolderAccessRole.EDITOR: EDITOR_PERMISSIONS,
+    FolderAccessRole.OWNER: OWNER_PERMISSIONS,
+}
+
+
+def _get_permissions_from_role(role: FolderAccessRole) -> _FolderPermissions:
+    return _ROLE_TO_PERMISSIONS[role]
+
+
+def _requires(*permissions: _FolderPermissions) -> _FolderPermissions:
+    if len(permissions) == 0:
+        return _make_permissions()
+    return _or_dicts_list(permissions)
+
+
+def _get_true_permissions(
+    permissions: _FolderPermissions, table
+) -> ColumnElement | bool:
+    """compose SQL where clause where only for the entries that are True"""
+    clauses: list[ColumnElement] = []
+
+    if permissions.read:
+        clauses.append(table.c.read.is_(True))
+    if permissions.write:
+        clauses.append(table.c.write.is_(True))
+    if permissions.delete:
+        clauses.append(table.c.delete.is_(True))
+
+    return sa.and_(*clauses) if clauses else True
+
+
+def _get_all_permissions(permissions: _FolderPermissions, table) -> ColumnElement:
+    return sa.and_(
+        table.c.read.is_(permissions.read),
+        table.c.write.is_(permissions.write),
+        table.c.delete.is_(permissions.delete),
+    )
+
+
+###
+### UTILS
+###
+
+
+class FolderName(ConstrainedStr):
+    regex = re.compile(
+        r'^(?!.*[<>:"/\\|?*\]])(?!.*\b(?:LPT9|COM1|LPT1|COM2|LPT3|LPT4|CON|COM5|COM3|COM4|AUX|PRN|LPT2|LPT5|COM6|LPT7|NUL|COM8|LPT6|COM9|COM7|LPT8)\b).+$',
```

Comment:
this is nice.. but it is a bit crazy. LPT is the parallel port right? and COM the serial port... what about USB and whatknot?
I guess we should define restrictions once we get there no? cause I guess this will never cover it all.